### PR TITLE
Feature/add sort to users

### PR DESCRIPTION
**Pull request by AWitcherONS**
### What

User Requst to be able to sort users 

### How to review

Code should conform to current standards 
Unit tests 
Component tests
and Swagger

### Who can review

Anyone but me

### CODE REVIEW

File: api/users_test.go [hunk 1]
```diff
@@ -25,6 +25,9 @@ const usersEndPointWithActiveFilterTrue = "http://localhost:25600/v1/users?activ
 const usersEndPointWithActiveFilterFalse = "http://localhost:25600/v1/users?active=false"
 const usersEndPointWithActiveFilterError = "http://localhost:25600/v1/user?active=false"
 
+const usersEndPointWithSortByEmail = "http://localhost:25600/v1/users?sort=email"
+const usersEndPointWithSortByEmailAcs = "http://localhost:25600/v1/users?sort=email:asc"
```
Comment (lindenmckenzie) [inline api/users_test.go]:
```suggestion
const usersEndPointWithSortByEmailAsc = "http://localhost:25600/v1/users?sort=email:asc"
```

File: features/steps/steps.go [hunk 2]
```diff
@@ -279,3 +287,24 @@ func (c *IdentityComponent) theResponseHeaderShouldContain(key, value string) (e
...
+	}
+	fmt.Println("\n\t--- actual ----", resBody)
+	fmt.Println("\n\t--- expected ----", expected)
+
+	if err = json.Unmarshal(resBody, &actual); err != nil {
+		return err
+	}
+
+	return nil
+}
```
Comment (lindenmckenzie) [inline features/steps/steps.go]: do we actually need a separate function here? Could we just use the same step description and point to the normal 'receive the following json' response?

File: features/users.feature [hunk 3]
```diff
@@ -269,6 +269,334 @@ Feature: Users
...
+        And a user with forename "Mary", lastname "Martin", email "email7@ons.gov.uk", id "id_3" and password "Passw0rd!" exists in the database
+        And I am an admin user
+        When I GET "/v1/users?sort=id"
+        Then I should receive the following JSON response with status "200":
+            """
+            {
+              "PaginationToken": "",
+              "count": 3,
+              "users": [
+                              {
```
Comment (lindenmckenzie) [inline features/users.feature]: indentation.

File: features/steps/steps.go (hunk 2 above)
Comment (lindenmckenzie) [reply features/steps/steps.go]: Looking at the feature file, I don't think you're actually using this.

File: swagger.yaml [hunk 4]
```diff
@@ -147,6 +147,10 @@ paths:
           name: active
           type: boolean
           description: filter on user active state.
+        - in: query
+          name: sort
+          type: string
+          description: sort the response by direction given key:asc or key:desc default asc, comma separates multi-field sorts
```
Comment (lindenmckenzie) [inline swagger.yaml]: could we note the default here as well please?

File: features/users.feature [hunk 5]
```diff
@@ -269,6 +269,334 @@ Feature: Users
...
+            }
+            """
+
+    @get-users-list
+    Scenario: GET /v1/users and checking the response status 200 with sort by id:desc
+        Given a user with forename "Adam", lastname "Adams", email "email5@ons.gov.uk", id "id_2" and password "Passw0rd!" exists in the database
+        And a user with forename "William", lastname "Williams", email "email9@ons.gov.uk", id "id_1" and password "Passw0rd!" exists in the database
+        And a user with forename "Mary", lastname "Martin", email "email7@ons.gov.uk", id "id_3" and password "Passw0rd!" exists in the database
+        And I am an admin user
+        When I GET "/v1/users?sort=id:asc"
```
Comment (lindenmckenzie) [inline features/users.feature]: Think this should be `:desc` from the scenario description. 

Which means the json needs to be inverted.

File: features/users.feature [hunk 6]
```diff
@@ -269,6 +269,334 @@ Feature: Users
...
+                  "forename": "William",
+                  "groups": [],
+                  "id": "id_5",
+                  "lastname": "Adams",
+                  "status": "CONFIRMED",
+                  "status_notes": ""
+                }
+                ]
+            }
+            """
```
Comment (lindenmckenzie) [inline features/users.feature:611]: what happens if someone puts in:

`/users?sort=dogs:asc`

File: query/sortUser.go [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+		case less(p, q):
+			return true
+		case less(q, p):
+			return false
+		}
+	}
+	return ms.less[k](p, q)
+}
+
+func SortBy(paramsSlice string, arr []models.UserParams) {
```
Comment (lindenmckenzie) [inline query/sortUser.go]: I'm not sure about how descriptive the 'paramsSlice' is. 

Godoc would be useful here too (and on the above docs)

File: features/steps/steps.go [hunk 8]
```diff
@@ -46,6 +47,8 @@ func (c *IdentityComponent) RegisterSteps(ctx *godog.ScenarioContext) {
 	ctx.Step(`^request header Accept is "([^"]*)"$`, c.requestHeaderAcceptIs)
 	ctx.Step(`^the response should match the following csv:$`, c.theResponseShouldMatchTheFollowingCsv)
 	ctx.Step(`^the response header "([^"]*)" should contain "([^"]*)"$`, c.theResponseHeaderShouldContain)
+	ctx.Step(`^I should receive the following sorted JSON response with status "([^"]*)":$`, c.iShouldReceiveTheFollowingSortedJSONResponseWithStatus)
```
Comment (lindenmckenzie) [inline features/steps/steps.go]: Not sure this is being used in the feature file.

File: query/sortUser.go [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+	emailDesc := func(c1, c2 *models.UserParams) bool { return c1.Email > c2.Email }
+	idAsc := func(c1, c2 *models.UserParams) bool { return c1.ID < c2.ID }
+	idDesc := func(c1, c2 *models.UserParams) bool { return c1.ID > c2.ID }
+
+	p1 := strings.Split(paramsSlice, ",")
+
+	var orderFunc []lessFunc
+	for _, p2 := range p1 {
+		p3 := strings.Split(p2, ":")
+		if len(p3) > 1 {
```
Comment (lindenmckenzie) [inline query/sortUser.go]: I think these `px` variables could do with some more descriptive names.

File: query/sortUser.go [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+
+func SortBy(paramsSlice string, arr []models.UserParams) {
+	forenameAsc := func(c1, c2 *models.UserParams) bool { return c1.Forename < c2.Forename }
+	forenameDesc := func(c1, c2 *models.UserParams) bool { return c1.Forename > c2.Forename }
+	lastnameAsc := func(c1, c2 *models.UserParams) bool { return c1.Lastname < c2.Lastname }
+	lastnameDesc := func(c1, c2 *models.UserParams) bool { return c1.Lastname > c2.Lastname }
+	emailAsc := func(c1, c2 *models.UserParams) bool { return c1.Email < c2.Email }
+	emailDesc := func(c1, c2 *models.UserParams) bool { return c1.Email > c2.Email }
+	idAsc := func(c1, c2 *models.UserParams) bool { return c1.ID < c2.ID }
+	idDesc := func(c1, c2 *models.UserParams) bool { return c1.ID > c2.ID }
```
Comment (lindenmckenzie) [inline query/sortUser.go]: Could this be made more generic by passing attributes in? i.e. we could map the incoming attributes to the underlying struct variables. Covers a bit of those nested if statements later as well.

File: swagger.yaml (hunk 4 above)
Comment (lindenmckenzie) [reply swagger.yaml]: an enum of all the options would be good too.

**Review by lindenmckenzie — COMMENTED** [review]
Some comments - happy to have a longer discussion on it tomorrow in the office if you're in :)

File: api/users_test.go (hunk 1 above)
Comment (AWitcherONS) [reply api/users_test.go:29]: :asc is not mandatory however :desc is mandatory 
 I am testing both

File: features/users.feature (hunk 6 above)
Comment (AWitcherONS) [reply features/users.feature:611]: Good test ... it shouldn't fail just will not sort the query

File: api/users_test.go [hunk 11]
```diff
@@ -24,7 +25,11 @@ const usersEndPoint = "http://localhost:25600/v1/users"
 const usersEndPointWithActiveFilterTrue = "http://localhost:25600/v1/users?active=true"
 const usersEndPointWithActiveFilterFalse = "http://localhost:25600/v1/users?active=false"
 const usersEndPointWithActiveFilterError = "http://localhost:25600/v1/user?active=false"
-
+const usersEndPointWithSortByEmail = "http://localhost:25600/v1/users?sort=email"
+const usersEndPointWithSortByEmailAsc = "http://localhost:25600/v1/users?sort=email:asc"
+const usersEndPointWithSortByEmailDesc = "http://localhost:25600/v1/users?sort=email:desc"
+const usersEndPointWithSortBy2fieldsDesc = "http://localhost:25600/v1/users?sort=forename:desc,lastname:desc"
+const usersEndPointWithSortBy2knowFieldsandUnknown = "http://localhost:25600/v1/users?sort=forename:desc,lastname:desc,dog"
```
Comment (lindenmckenzie) [inline api/users_test.go]:
```suggestion
const usersEndPointWithSortBy2KnownFieldsAndUnknown = "http://localhost:25600/v1/users?sort=forename:desc,lastname:desc,dog"
```

File: api/users_test.go [hunk 12]
```diff
@@ -24,7 +25,11 @@ const usersEndPoint = "http://localhost:25600/v1/users"
 const usersEndPointWithActiveFilterTrue = "http://localhost:25600/v1/users?active=true"
 const usersEndPointWithActiveFilterFalse = "http://localhost:25600/v1/users?active=false"
 const usersEndPointWithActiveFilterError = "http://localhost:25600/v1/user?active=false"
-
+const usersEndPointWithSortByEmail = "http://localhost:25600/v1/users?sort=email"
+const usersEndPointWithSortByEmailAsc = "http://localhost:25600/v1/users?sort=email:asc"
+const usersEndPointWithSortByEmailDesc = "http://localhost:25600/v1/users?sort=email:desc"
+const usersEndPointWithSortBy2fieldsDesc = "http://localhost:25600/v1/users?sort=forename:desc,lastname:desc"
```
Comment (lindenmckenzie) [inline api/users_test.go]:
```suggestion
const usersEndPointWithSortBy2FieldsDesc = "http://localhost:25600/v1/users?sort=forename:desc,lastname:desc"
```

File: swagger.yaml [hunk 13]
```diff
@@ -697,7 +718,6 @@ paths:
           required: false
           default: created
           enum:
-            - created
```
Comment (lindenmckenzie) [inline swagger.yaml]: this is being removed from get groups?

File: swagger.yaml [hunk 14]
```diff
@@ -147,6 +147,27 @@ paths:
           name: active
           type: boolean
           description: filter on user active state.
+        - in: query
+          name: sort
+          type: string
```
Comment (lindenmckenzie) [inline swagger.yaml:152]:
```suggestion
          type: string
          default: id:asc
```

**Review by lindenmckenzie — COMMENTED** [review]
Some trivial comments.

**Review by lindenmckenzie — APPROVED** [review]

**Pull request merged**